cli/src/host_perf_check.rs

```diff
@@ -0,0 +1,145 @@
+		Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(false),
+		Err(err) => return Err(err),
+	};
+	let mut reader = io::BufReader::new(file);
+
+	reader.read_exact(&mut buf)?;
+
+	Ok(host_name == buf)
+}
+
+fn save_check_passed_file(path: &Path) -> io::Result<()> {
+	let host_name_max_len = unistd::SysconfVar::HOST_NAME_MAX as usize;
+	let mut host_name = vec![0u8; host_name_max_len];
+	unistd::gethostname(&mut host_name).map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+
+	let mut file = OpenOptions::new().truncate(true).create(true).write(true).open(path)?;
+
+	file.write(&host_name)?;
+
+	Ok(())
+}
+
+pub fn host_perf_check(result_cache_path: &Path) -> Result<(), PerfCheckError> {
+	const CODE_SIZE_LIMIT: usize = 1024usize.pow(3);
+	const CHECK_PASSED_FILE_NAME: &str = ".perf_check_passed";
+	let wasm_code = kusama_runtime::WASM_BINARY.ok_or(PerfCheckError::WasmBinaryMissing)?;
+
+	let check_passed_file_path = result_cache_path.join(CHECK_PASSED_FILE_NAME);
+
+	if let Ok(true) = is_perf_check_done(&check_passed_file_path) {
+		info!("Performance check skipped: already passed (cached at {:?})", check_passed_file_path);
+		return Ok(())
+	}
+
+	// Decompress the code before running checks.
+	let code = sp_maybe_compressed_blob::decompress(wasm_code, CODE_SIZE_LIMIT)
+		.or(Err(PerfCheckError::CodeDecompressionFailed))?;
+
+	info!("Running the performance checks...");
+
+	pvf_perf_check(code.as_ref())?;
+
+	erasure_coding_perf_check(code.as_ref())?;
+
+	// Persist successful result.
+	if let Err(err) = save_check_passed_file(&check_passed_file_path) {
+		info!("Couldn't persist check result at {:?}: {}", check_passed_file_path, err.to_string());
+	}
+
+	Ok(())
+}
+
+fn green_threshold(duration: Duration) -> Duration {
+	duration * 4 / 5
+}
+
+fn perf_check(
+	test_name: &str,
+	time_limit: Duration,
+	test: impl Fn() -> Result<(), PerfCheckError>,
+) -> Result<(), PerfCheckError> {
+	let start = Instant::now();
+
+	test()?;
+
+	let elapsed = start.elapsed();
+	if elapsed < green_threshold(time_limit) {
+		info!("🟢 {} performance check passed, elapsed: {:?}", test_name, elapsed);
+		Ok(())
+	} else if elapsed <= time_limit {
+		info!(
+			"🟡 {} performance check passed, {:?} limit almost exceeded, elapsed: {:?}",
+			test_name, time_limit, elapsed
+		);
+		Ok(())
+	} else {
+		Err(PerfCheckError::TimeOut { elapsed, limit: time_limit })
+	}
+}
+
+fn pvf_perf_check(code: &[u8]) -> Result<(), PerfCheckError> {
+	const PREPARE_TIME_LIMIT: Duration = Duration::from_secs(20);
+
+	perf_check("PVF-prepare", PREPARE_TIME_LIMIT, || {
+		// Recreate the pipeline from the pvf prepare worker.
+		let blob =
+			polkadot_node_core_pvf::prevalidate(code.as_ref()).map_err(PerfCheckError::from)?;
+		polkadot_node_core_pvf::prepare(blob).map_err(PerfCheckError::from)?;
+		Ok(())
+	})
+}
+
+fn erasure_coding_perf_check(data: &[u8]) -> Result<(), PerfCheckError> {
+	const ERASURE_CODING_TIME_LIMIT: Duration = Duration::from_secs(1);
```

Comment:
Where did you get these values from?

Reply:
I don't have a reference machine to run this on, that's why I put "random" timeouts. Do you mean we should obtain these values similar to benchmarks ran by a bot?

Reply:
Exactly.